Initial intercommunalities support

noirbizarre opened the pull request:
This PR adds initial intercommunalities support.

This is a first iteration lacking:
- some naming rules (see tests)
- `sources/epci/fusions.csv` completion
- handle merges handling from `sources/epci/fusions.csv`
- intermediate states generation (ie. town changes during the year)
- population computation from towns `pmun` instead of the provided `ptot`


## Sources

Sources are extracted downloaded and reformatted from "[Listes et compositions des EPCI à fiscalité propre publiées par la Direction Générale des collectivités locales](https://www.collectivites-locales.gouv.fr/liste-et-composition-des-epci-a-fiscalite-propre)".

There is both a french and english documentation for thoses but some sections still needs to be written (there is some TODOs)

There is an intermediate `EPCI.ods` with an embedded python macro to help the formatting
and the generation of the CSVs in a proper format.

```bash
$ wc -l sources/epci/????.csv
   19129 sources/epci/1999.csv
   21348 sources/epci/2000.csv
   23498 sources/epci/2001.csv
   26871 sources/epci/2002.csv
   29755 sources/epci/2003.csv
   31429 sources/epci/2004.csv
   32309 sources/epci/2005.csv
   32924 sources/epci/2006.csv
   33414 sources/epci/2007.csv
   33639 sources/epci/2008.csv
   34167 sources/epci/2009.csv
   34775 sources/epci/2010.csv
   35042 sources/epci/2011.csv
   35306 sources/epci/2012.csv
   36050 sources/epci/2013.csv
   36615 sources/epci/2014.csv
   36589 sources/epci/2015.csv
   35859 sources/epci/2016.csv
   35412 sources/epci/2017.csv
  604131 total
```

Uniformized sources will be published back on data.gouv.fr

**Note**: Sources are sorted on the `siren` field. This is very important because the processing rely on it.

## Processing

Processing takes about 3 hours and a half and represents 604112 calls to `Towns.valid_at()` which explain the duration.
In consequence, the intercommunalities processing has been made optionnal with the `--intercommunalities` flag.

Names format changes from one year to another and there is some typos, this is why we try to fix and normalize each name. The normalization process still needs some improvement on (between others):
- rules for St/Ste/Saint/Sainte
- rules for Hts/Ht/Hte/Hauts/Haut/Haute
- rules for "interdepartementale", "interdépartementale", "interrégionale", "Intercommunale"
- some dashes and spaces normalization (ie. "Caux-fleur de Lin" vs. "Caux - Fleur de Lin")
- some accents normalization/replacements (ie. "ile/Ile" vs. "île/Île")
In the meantime, this produces more changes than reality (name changes).

Sometimes, an acronym is provided in the name, between parenthesis or brackets. We try to properly extract and keep it when possible. This seems to be a valuable information because some people only know the EPCI acronym instead of the full name.


## Export

The export is fully documented into `exports/epci` keeping the format of others exports.


## Extra additions

This PR also adds some logging in order to see the progress of processing.
The logging is handled by [click-log](https://click-log.readthedocs.io/) which provides sane defaults for click integration.
This adds an extra `-v/--verbosity LVL` option where `LVL` is one of `CRITICAL`, `ERROR`, `WARNING`, `INFO` or `DEBUG` (case does not matter) and defaults to `INFO`.

## Result

The last execution produced the following on a core i7 with 16Gb RAM:
```bash
$ time python -m geohisto --intercommunalities -v debug && wc -l exports/epci/epci.csv
Loading towns
Loading history
Loading populations
Loading counties
Computing history from actions
Apply special cases
Compute ancestors
Update the parents for each town.
Processing intercommunalities from sources/epci (1999-2017)
debug: Load intercommunalities from sources/epci/1999.csv
debug: Load intercommunalities from sources/epci/2000.csv
debug: Load intercommunalities from sources/epci/2001.csv
debug: Load intercommunalities from sources/epci/2002.csv
debug: Load intercommunalities from sources/epci/2003.csv
debug: Load intercommunalities from sources/epci/2004.csv
debug: Load intercommunalities from sources/epci/2005.csv
debug: Load intercommunalities from sources/epci/2006.csv
debug: Load intercommunalities from sources/epci/2007.csv
debug: Load intercommunalities from sources/epci/2008.csv
debug: Load intercommunalities from sources/epci/2009.csv
debug: Load intercommunalities from sources/epci/2010.csv
debug: Load intercommunalities from sources/epci/2011.csv
debug: Load intercommunalities from sources/epci/2012.csv
debug: Load intercommunalities from sources/epci/2013.csv
debug: Load intercommunalities from sources/epci/2014.csv
debug: Load intercommunalities from sources/epci/2015.csv
debug: Load intercommunalities from sources/epci/2016.csv
debug: Load intercommunalities from sources/epci/2017.csv
Writing towns file to exports/communes/communes.csv
Writing towns head file to exports/communes/communes_head.csv
Writing intercommunalities file to exports/epci/epci.csv

real	215m2,377s
user	212m53,826s
sys	0m15,816s

43244 exports/epci/epci.csv
```

## Possible improvements

- Allows to process intercommunalities separately by loading the towns export
- Improve the EPCI.ods macro to automatically sort on the `siren` field before exporting
- Implement any of the described lacks


## Note
Sorry for the added size to the repository but I can't see any other option thant providing the source and export files

davidbgk commented:
> Sorry for the added size to the repository but I can't see any other option thant providing the source and export files

Maybe at least split the PR with a first commit which only adds sources?

davidbgk commented:
How can I review the `EPCI.ods` file? I don't get the necessity for it?

davidbgk commented:
Can you try to add a [`lru_cache`](https://docs.python.org/dev/library/functools.html#functools.lru_cache) on `valid_at` function to see if it reduces time? (let's say with a `max_size=None` at first 😉  )

noirbizarre commented:
1. I will :+1: 
2. there is no change at all on the town generation, the diff is only messed because of the factorising of `Collections` and `Items` but methods are left unchanged
3. files loading does not takes much time, this is the `Town.valid_at()` querying which takes a lot of time (written in the PR). Splitting this in 2 steps means of the following:
    - storing 604112 objects in memory plus the towns
    - writing and reloading the intermediate step (which means taking the same amount of time)

  Right now, the process avoid that by using a generator

> Maybe at least split the PR with a first commit which only adds sources?

I don't see the point but I will

> How can I review the EPCI.ods file? I don't get the necessity for it?

Fetch the branch and open it in LibreOffice or OpenOffice.
Just try to reformat and serialize 19 CSV files using a common format (multiple times), also perform some search in the files and you will see the necessity ;)
To me it was a huge gains of time and comfort

> Can you try to add a lru_cache on valid_at

I don't think this will do anything more than load the memory and introduce overcoast: each town is looked once only for a given year because each town can only be in a single intercommunality at once. :cry:

davidbgk commented:
> each town is looked once only for a given year because each town can only be in a single intercommunality at once.

Crap :(

What about iterating on towns an injecting those into given EPCIs? If there are fewer ones that might be faster.

noirbizarre commented:
This is even worse: there is less lines in each intercommunality year than towns but iterating towns means iterating each files (except years after a given validity) the number of towns and this generate event more iteration (I've already examined this possibility) :(

davidbgk commented on geohisto/__main__.py:
```diff
@@ -29,17 +36,29 @@ def main(at_date):
     compute_populations(populations, towns)
     compute_parents(counties, towns)
+    if intercommunalities:
+        intercommunalities = load_intercommunalities(towns)
```
Update the comment about the order if intercommunalities must be loaded (computed?) last/after towns.

davidbgk commented on geohisto/constants.py:
```diff
@@ -87,3 +87,13 @@
+END_REMOVED = 'removed'
+END_MERGE = 'merged'
+
+INTERCOMMUNALITY_START_YEAR = 1999
```
Use `INTERCOMMUNALITY_START_DATE.year` in your code.

noirbizarre replied:
Done

davidbgk commented on geohisto/intercommunalities.py:
```diff
@@ -0,0 +1,287 @@
+}
+
+
+def should_replace_kind(name, kind):
```
Pretty intense 😅 maybe add a comment about why?

noirbizarre replied:
done

davidbgk commented on geohisto/intercommunalities.py:
```diff
@@ -0,0 +1,287 @@
+
+
+def fix_casing(name):
+    return ' '.join(fix_word_casing(word) for word in name.split(' '))
```
I think the parameter of `split` is not required here.

davidbgk commented on geohisto/intercommunalities.py:
```diff
@@ -0,0 +1,287 @@
+    kind = line['nature']
+    name = fix_typos(name)
+    name = ' '.join(name.split())  # replace multiple spaces
+    if name.isupper():
```
Testing that before the split/join would allow to directly pass the list to `fix_casing`

noirbizarre replied:
Good idea, done, but I don't know why it seems slower (test execution take `0.05 seconds` instead of `0.04 seconds` before the change)

davidbgk commented on geohisto/intercommunalities.py:
```diff
@@ -0,0 +1,287 @@
+    elif name.lower().startswith(tuple(
+            ' '.join((kind.lower(), prefix, prep)) for prep in PREPOSITIONS
+            for prefix in COMMON_PREFIXES
+            )):
```
Bad indent. I would put the parenthesis at the end of the line above.

noirbizarre replied:
Changed, but the files were already linted and linter was not complaining about this one.
Both are acceptable. Personnaly I prefer the version before the change, easier to read the 2 conditions.

davidbgk commented on geohisto/intercommunalities.py:
```diff
@@ -0,0 +1,287 @@
+    if not name.startswith('Tour'):
+        name = re.sub(RE_WITH_PRENTHESIS, '', name).strip()
+    # First char is always upper
+    return name[0].upper() + name[1:]
```
`name.capitalize()`?

noirbizarre replied:
Nop, `capitalize()` put all the end of the string in lower case which is not the desired behavior

davidbgk commented on geohisto/intercommunalities.py:
```diff
@@ -0,0 +1,287 @@
+        return match.group(1).strip().upper() if match else None
+
+
+def siren_key(line):
```
It looks weird to have a method just for that compared to others.

davidbgk commented on geohisto/intercommunalities.py:
```diff
@@ -0,0 +1,287 @@
+    with open(filename) as epci_csv:
+        all_lines = csv.DictReader(epci_csv, delimiter=';', quotechar='"')
+        # No need to sort CSV are already sorted by SIREN
+        # all_lines = sorted(all_lines, key=siren_key)
```
Remove the line (not the comment)

davidbgk commented on geohisto/intercommunalities.py:
```diff
@@ -0,0 +1,287 @@
+        # No need to sort CSV are already sorted by SIREN
+        # all_lines = sorted(all_lines, key=siren_key)
+
+        for siren, lines in groupby(all_lines, siren_key):
```
Ah, I get it, well maybe a lambda candidate in that case?

davidbgk commented on geohisto/intercommunalities.py:
```diff
@@ -0,0 +1,287 @@
+def attach_town(intercommunality, insee, validity, towns):
+    try:
+        intercommunality.towns.add(towns.get_current(insee, validity).id)
+    except:
```
Please put a specific exception here.

davidbgk commented on geohisto/intercommunalities.py:
```diff
@@ -0,0 +1,287 @@
+    try:
+        intercommunality.towns.add(towns.get_current(insee, validity).id)
+    except:
+        log.error('Failed for %s on %s@%s',
```
Curious to know which ones are failing, potential bugs within towns?

noirbizarre replied:
See https://github.com/etalab/geohisto/pull/53#issuecomment-323022073

davidbgk commented on geohisto/intercommunalities.py:
```diff
@@ -0,0 +1,287 @@
+)
+
+
+def write_intercommunalities_on(filename, intercommunalities,
```
Moving it to `exports.py` will ease future refactoring.

noirbizarre replied:
Not sure about it. It was before but it was the only part of intercommunalities process split out in another module so I grouped it here so the whole intercommunalities part is isolated.

davidbgk commented on geohisto/models.py:
```diff
@@ -1,73 +1,70 @@
-        is_created = town.id not in self
-        self[town.id] = town
+
+class Collection(object):
```
No need to inherit from `object`?

davidbgk replied:
I think it should inherit from `dict` or be named `CollectionMixin` and documented to be part of a class implementing the `dict` interface somehow.

noirbizarre replied:
Inheriting from dict is not possible as Towns inherit from OrderedDict.

davidbgk commented on geohisto/models.py:
```diff
@@ -90,8 +87,71 @@ def get_current(self, depcom, valid_datetime):
+            self.replace_successor(from_town, town)
+
+
+class Item(object):
```
Same here, it uses the namedtuple interface internally (`._replace`, such an ugly one btw…)

noirbizarre replied:
Yes but `namedtuple` is not factorizable: it not a `class` but a `class` factory and fields differs between `Town` and `Intercommunality`

noirbizarre replied:
`._replace` is bad because it use the private `_`-naming convention but it's fully documented and these classes exposes upper level methods.

davidbgk commented on geohisto/models.py:
```diff
@@ -167,57 +227,127 @@ def add_modification(self, modification):
-        return self._replace(**{'successors': successors})
+    def valid_at(self, valid_date, siren=None):
+        """
+        Return a list of Intercommunalities existing at the given `valid_date`.
```
Line length :)

noirbizarre replied:
Checked this is OK

davidbgk commented on geohisto/models.py:
```diff
@@ -167,57 +227,127 @@ def add_modification(self, modification):
+            _items = self.values()
+        return [item for item in _items if item.valid_at(valid_date)]
+    #
+    # def get_current(self, siren, valid_date):
```
Remove if not in use.

noirbizarre replied:
Already removed, just not pushed yet

davidbgk commented on geohisto/models.py:
```diff
@@ -167,57 +227,127 @@ def add_modification(self, modification):
+        changes = []
+        if intercommunality.towns != current.towns:
+            # This is a perimeter change
+            changes.append(END_INNER)
```
What about naming that const `INTERCOMMUNALITY_PERIMETER_CHANGE`? (general comment: `epci` is shorter that `intercommunality` 😅 )

Change below to be consistent if you agree.

noirbizarre replied:
Yes it is, but it doesn't exists in english ;)

noirbizarre replied:
`INTERCOMMUNALITY_PERIMETER_CHANGE` does not fit, it can simply be a component town renamed, merged or splitted... perimeter does not change is this case.
It took `INTERCOMMUNALITY_INNER_CHANGE`

davidbgk commented on geohisto/models.py:
```diff
@@ -167,57 +227,127 @@ def add_modification(self, modification):
+        return False
+
+
+class Intercommunality(Item, namedtuple('Intercommunalitite', [
```
`Intercommunalitite` should be `Intercommunality` 👼

davidbgk commented on geohisto/models.py:
```diff
@@ -167,57 +227,127 @@ def add_modification(self, modification):
+        'end_reason',  # Reason this intercommunality is ended. See above.
+        'taxmodel',  # One of 4TX, TPU, FA, FPU
+        'population',  # Known population  from last census
+        ])):
```
Bad indent. Proposition:

```python
class Intercommunality(Item, namedtuple(
        'Intercommunalitite', [
            'id',  # GeoID
            'siren',  # SIREN identifier
            'name',  # Normalized name
            'acronym',  # Public acronym, not always initials
            'kind',  # One of CC, CA, CU, CV, DISTRICT, METRO, METRO69, SAN
            'successors',  # List of successors identifiers
            'ancestors',  # List of ancestors identifiers
            'towns',  # Set of components towns identifiers
            'missing_towns',  # Set of towns INSEE code not matched to a town
            'start_date',  # Start validity date
            'end_date',  # End validity date
            'end_reason',  # Reason this intercommunality is ended. See above.
            'taxmodel',  # One of 4TX, TPU, FA, FPU
            'population',  # Known population  from last census
        ])):
```

davidbgk commented on geohisto/models.py:
```diff
@@ -167,57 +227,127 @@ def add_modification(self, modification):
-            return False
-        return self.start_datetime <= valid_datetime <= self.end_datetime
+    def __new__(cls, **kwargs):
+        '''Set default values to allow shorter constructors'''
```
Good idea, I'll steal that for Towns too 👍

noirbizarre replied:
;)

noirbizarre commented on exports/epci/README.md:
```diff
@@ -0,0 +1,42 @@
+The `id` column is unique, the `siren` one is NOT.
+Arbitrarily, the far future end date has been set to `9999-12-31 23:59:59`.
+
+A new entry/line is registered when the intercommunality changed one of its:
```
Yes but only merge is not yet handled, but should be in another PR (because the `fusions.csv` needs to be filled)

noirbizarre commented on exports/epci/README.md:
```diff
@@ -0,0 +1,42 @@
+
+This is a regular CSV file with value separated by commas and a header line with previously described column names.
+
+You can have a look at the file using Github preview if you click on it.
```
Already done

noirbizarre commented on exports/epci/README.md:
```diff
@@ -0,0 +1,42 @@
+* `name`: The name of the intercommunality.
+* `acronym`: An optionnal known acronym or short name.
+* `kind`: The legal form of the intercommunality.
+* `taxmodel`: The tax model under which the intercommunality is.
```
I think so. From that depends citizen taxation

noirbizarre commented on README.md:
```diff
@@ -31,6 +31,14 @@ Optionally, you can specify a date to only export towns valid at that given date
 
 It will be generated within the `exports/communes/` folder with an explicit name.
 
+To also generate the intercommunalities, you need to add the `--intercommunalities` flag.
```
done

noirbizarre commented:
# Last executions

## Filter on year
```bash
$ time python -m geohisto --intercommunalities -v debug && wc -l exports/epci/epci.csv
Loading towns
Loading history
Loading populations
Loading counties
Computing history from actions
Applying special cases
Computing ancestors
Updating parents
Processing intercommunalities from sources/epci (1999-2017)
debug: Load intercommunalities from sources/epci/1999.csv
debug: Load intercommunalities from sources/epci/2000.csv
debug: Load intercommunalities from sources/epci/2001.csv
debug: Load intercommunalities from sources/epci/2002.csv
debug: Load intercommunalities from sources/epci/2003.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2003-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2004.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2004-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2005.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2005-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2006.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2006-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2007.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2007-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2008.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2008-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2009.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2009-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2010.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2010-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2011.csv
^[[1;2Derror: Failed for Communauté de Communes des Trois Rivières on 14697@2011-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2012.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2012-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2013.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2013-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2014.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2014-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2015.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2015-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2016.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2016-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2017.csv
Writing towns file to exports/communes/communes.csv
Writing exports/communes/communes.csv head file to exports/communes/communes_head.csv
Writing intercommunalities file to exports/epci/epci.csv
Writing exports/epci/epci.csv head file to exports/epci/epci_head.csv

real	258m12,440s
user	159m38,164s
sys	0m7,019s
43243 exports/epci/epci.csv
```

## Sorts removed
```bash
time python -m geohisto --intercommunalities -v debug && wc -l exports/epci/epci.csv
Loading towns
Loading history
Loading populations
Loading counties
Computing history from actions
Applying special cases
Computing ancestors
Updating parents
Processing intercommunalities from sources/epci (1999-2017)
debug: Load intercommunalities from sources/epci/1999.csv
debug: Load intercommunalities from sources/epci/2000.csv
debug: Load intercommunalities from sources/epci/2001.csv
debug: Load intercommunalities from sources/epci/2002.csv
debug: Load intercommunalities from sources/epci/2003.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2003-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2004.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2004-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2005.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2005-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2006.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2006-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2007.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2007-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2008.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2008-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2009.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2009-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2010.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2010-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2011.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2011-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2012.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2012-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2013.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2013-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2014.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2014-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2015.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2015-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2016.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2016-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2017.csv
Writing towns file to exports/communes/communes.csv
Writing exports/communes/communes.csv head file to exports/communes/communes_head.csv
Writing intercommunalities file to exports/epci/epci.csv
Writing exports/epci/epci.csv head file to exports/epci/epci_head.csv

real	154m38,616s
user	154m24,401s
sys	0m13,624s
40495 exports/epci/epci.csv
```

## Notes
Resulting line numbers differs because of some addition in `extract_names` (which means more convergence).

Removing the extra sorts and applying some tuning gives a gain of 1h !

Splitting on year introduce one missing town for INSEE code 14697 starting on year 2003.
This town is:
- https://www.insee.fr/fr/metadonnees/cog/commune/COM14697-oudon
- https://www.insee.fr/fr/metadonnees/cog/commune/COM14697-totes
I think this has something to do with https://github.com/etalab/geohisto/blob/master/geohisto/specials.py#L432-L459 which doesn't handle the `14697` INSEE code

davidbgk commented on geohisto/intercommunalities.py:
```diff
@@ -0,0 +1,305 @@
+        return word.capitalize()
+
+
+def fix_casing(name):
```
Not anymore in use

noirbizarre commented:
I can remove the EPCI.ods from the PR and only submit it on Data.gouv.fr (just an idea)

davidbgk commented:
Before I forget, it deserves a 10.0.0 within the CHANGELOG 🎉

davidbgk commented:
> I can remove the EPCI.ods from the PR and only submit it on Data.gouv.fr (just an idea)

The problem with that file is that it cannot be reviewed nor display diffs on update so I'm in favor of having it as an external resource but I'm still not sure of the usage. Is it useful for input data sources or to deal with the exported one? Can I have a 5 min demo tomorrow?

davidbgk commented:
> Removing the extra sorts and applying some tuning gives a gain of 1h !

Not bad 😅 

> Splitting on year introduce one missing town for INSEE code 14697 starting on year 2003.

Alright, this is probably a bug on towns’ side, cases handled as ”specials” are manual fixes. Can you fill an issue about that?

davidbgk commented on README.md:
```diff
@@ -31,6 +33,14 @@ Optionally, you can specify a date to only export towns valid at that given date
+
+    $ python -m geohisto --intercommunalities
+
+The whole process takes about 3 hours to generate both towns and intercommunalities exports.
```
2 hours now?

davidbgk commented on geohisto/models.py:
```diff
@@ -1,84 +1,88 @@
-        _towns.sort()  # Useful for tests.
-        return _towns
+                  if getattr(item, k) == v]
+        if sort:
```
Idea: do not sort at all here but sort at the end of the `compute()` method once and for all. It looks cleaner to me and should not affect tests (if that's still the case, let's fix that at the tests level!).

noirbizarre replied:
I will :+1:

davidbgk commented on geohisto/models.py:
```diff
@@ -167,57 +237,140 @@ def add_modification(self, modification):
+    def latest(self, siren):
+        """Get the latest valid intercommunality for a given `siren`."""
+        # No need to sort, there is theoricaly only one town
+        # with a gvien siren at a time
```
given*

noirbizarre commented:
> > Splitting on year introduce one missing town for INSEE code 14697 starting on year 2003.

> Alright, this is probably a bug on towns’ side, cases handled as ”specials” are manual fixes. Can you fill an issue about that?

Done: #54 

> Before I forget, it deserves a 10.0.0 within the CHANGELOG :tada:

Added in the last commit :+1: 

> > I can remove the EPCI.ods from the PR and only submit it on Data.gouv.fr (just an idea)

> The problem with that file is that it cannot be reviewed nor display diffs on update so I'm in favor of having it as an external resource but I'm still not sure of the usage. Is it useful for input data sources or to deal with the exported one? Can I have a 5 min demo tomorrow?

Will be published with the CSVs on Data.gouv.fr [here](https://www.data.gouv.fr/fr/datasets/liste-et-composition-des-epci-a-fiscalite-propre/) (private right now) and removed from the PR. The macro has been splitted out and published as [a gist](https://gist.github.com/noirbizarre/27533d83556fd010c13c994604e158a2)

I can do a five minute demo but there is nothing much to demonstrate. This allows to:
- have all files properly encoded loaded once
- sort on the SIREN field for proper export 
- adds merges and legend data
- search and filter on names and SIREN to investigate issues
- easily export all the sheets once and with the same paremeters with the provided macro

noirbizarre commented:
Diff implemented and improved  (see last commits)

## Tests duration
Initialy: 304.52 seconds
Diff applyied (Sort once at actions.compute): 319.55 seconds
Use generators: 224.33 seconds
Tune actions.compute sort: 210.43 seconds

## Full execution with changes
```
time python -m geohisto --intercommunalities -v debug && wc -l exports/epci/epci.csv
Loading towns
Loading history
Loading populations
Loading counties
Computing history from actions
Applying special cases
Computing ancestors
Updating parents
Processing intercommunalities from sources/epci (1999-2017)
debug: Load intercommunalities from sources/epci/1999.csv
debug: Load intercommunalities from sources/epci/2000.csv
debug: Load intercommunalities from sources/epci/2001.csv
debug: Load intercommunalities from sources/epci/2002.csv
debug: Load intercommunalities from sources/epci/2003.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2003-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2004.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2004-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2005.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2005-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2006.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2006-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2007.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2007-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2008.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2008-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2009.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2009-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2010.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2010-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2011.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2011-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2012.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2012-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2013.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2013-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2014.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2014-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2015.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2015-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2016.csv
error: Failed for Communauté de Communes des Trois Rivières on 14697@2016-01-01T00:00:00
debug: Load intercommunalities from sources/epci/2017.csv
Writing towns file to exports/communes/communes.csv
Writing exports/communes/communes.csv head file to exports/communes/communes_head.csv
Writing intercommunalities file to exports/epci/epci.csv
Writing exports/epci/epci.csv head file to exports/epci/epci_head.csv

real	84m21,209s
user	83m18,115s
sys	0m6,150s
40495 exports/epci/epci.csv
```

Obviously generators made the job !

There seams to me some more lines in exports, see last commit diff.
I think this fixed some edge cases (the few cases I checked are not false positive)

PS: if you have some addition, please provide an easily reusable format (commit, PR to my PR...), not a copy-pasted diff

davidbgk commented on geohisto/actions.py:
```diff
@@ -5,6 +5,8 @@
 """
 import logging
 
+from collections import OrderedDict
```
Not needed anymore.

davidbgk commented on exports/communes/communes_head.csv:
```diff
@@ -3,6 +3,7 @@ fr:commune:01001@1942-01-01,01001,1942-01-01 00:00:00,9999-12-31 23:59:59,L'Aber
 fr:commune:01002@1942-01-01,01002,1942-01-01 00:00:00,9999-12-31 23:59:59,L'Abergement-de-Varey,,,fr:departement:01@1860-07-01,236,0
 fr:commune:01003@1942-01-01,01003,1942-01-01 00:00:00,1973-12-31 23:59:59,Amareins,fr:commune:01165@1974-01-01,,fr:departement:01@1860-07-01,NULL,330
 fr:commune:01004@1942-01-01,01004,1942-01-01 00:00:00,1955-03-30 23:59:59,Ambérieu,fr:commune:01004@1955-03-31,,fr:departement:01@1860-07-01,NULL,100
+fr:commune:01004@1955-03-31,01004,1955-03-31 00:00:00,9999-12-31 23:59:59,Ambérieu-en-Bugey,,fr:commune:01004@1942-01-01,fr:departement:01@1860-07-01,14359,0
```
Sorting towns at the end allows to read more easily changes, I should have done this before!

noirbizarre replied:
Beneficial side effect :wink:

davidbgk commented on geohisto/actions.py:
```diff
@@ -3,6 +3,10 @@
 """
+import logging
+
+from collections import OrderedDict
```
Useless import
